packages/oidc-dpop-client-browser/src/dpop/tokenExchange.spec.ts

```diff
@@ -0,0 +1,430 @@
+/*
+ * Copyright 2020 Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import URL from "url-parse";
+import { IClient, IIssuerConfig } from "@inrupt/solid-client-authn-core";
+import { it, describe } from "@jest/globals";
+
+import { getTokens, TokenEndpointInput } from "./tokenExchange";
+import { decodeJwt, signJwt } from "./dpop";
+import { JWKECKey } from "jose";
+import { Response } from "cross-fetch";
+
+// Some spec-compliant claims are camel-cased.
+/* eslint-disable @typescript-eslint/camelcase */
+
+// The two following modules introduce randomness in the process, which prevents
+// making assumptions on the returned values. Mocking them out makes keys and
+// DPoP headers predictible.
+jest.mock("./keyGen", () => {
+  return {
+    generateJwkForDpop: (): JWKECKey => {
+      return {
+        kty: "EC",
+        kid: "oOArcXxcwvsaG21jAx_D5CHr4BgVCzCEtlfmNFQtU0s",
+        alg: "ES256",
+        crv: "P-256",
+        x: "0dGe_s-urLhD3mpqYqmSXrqUZApVV5ZNxMJXg7Vp-2A",
+        y: "-oMe9gGkpfIrnJ0aiSUHMdjqYVm5ZrGCeQmRKoIIfj8",
+        d: "yR1bCsR7m4hjFCvWo8Jw3OfNR4aiYDAFbBD9nkudJKM",
+      };
+    },
+  };
+});
+
+// Note that mockJwk and the mocked generateJwkForDpop from keygen return the
+// same mocked key, but due to jest initialisation, one cannot call the other.
```

Comment:
I think you might be able to call `jest.requireMock` in the mock initialisation function?

Reply:
Right, but what I meant is that I would like to do the following: 
```
const mockJwk = (): JWKECKey => {
  return {
    kty: "EC",
    kid: "oOArcXxcwvsaG21jAx_D5CHr4BgVCzCEtlfmNFQtU0s",
    alg: "ES256",
    crv: "P-256",
    x: "0dGe_s-urLhD3mpqYqmSXrqUZApVV5ZNxMJXg7Vp-2A",
    y: "-oMe9gGkpfIrnJ0aiSUHMdjqYVm5ZrGCeQmRKoIIfj8",
    d: "yR1bCsR7m4hjFCvWo8Jw3OfNR4aiYDAFbBD9nkudJKM",
  };
};
jest.mock("./keyGen", () => {
  return {
    generateJwkForDpop: mockJwk
});
```
but since `jest` hoists the mocked modules up, that results in `ReferenceError: Cannot access 'mockJwk' before initialization`. And the other, more "regular" issue is that generateJwkForDpop is `async`, which prevents doing `const mockJwk = generateJwkForDpop`, because `mockJwk` is used to initialize mocks at the top level (I think top-level `await` is on the wishlist for next ecmascript release, but not quite there yet).

Did you have another approach in mind ?

Reply:
Right, `mockJwk` is not defined when initialising the mocks, but I think it _is_ defined when functions defined in those mocks are executed. So you could do this, I think:

```
jest.mock("./keyGen", () => {
  return {
    generateJwkForDpop: () => mockJwk()
});
```

Reply:
:bulb: riiiight, very good point ! Fixed in f3b0e84f617039ea8ebf9c4a991a011765f2eb64